sdks/java/core/src/main/java/org/apache/beam/sdk/util/Reshuffle.java

```diff
@@ -55,28 +59,32 @@ private Reshuffle() {
     // If the input has already had its windows merged, then the GBK that performed the merge
     // will have set originalStrategy.getWindowFn() to InvalidWindows, causing the GBK contained
     // here to fail. Instead, we install a valid WindowFn that leaves all windows unchanged.
+    // Elements may not be shifted forwards in time by the GroupByKey, which requires updating the
+    // OutputTimeFn to output at the earliest encountered timestamp. Because this outputs as fast
+    // as possible, this should not hold the watermark.
     Window.Bound<KV<K, V>> rewindow =
-        Window.<KV<K, V>>into(
-                new IdentityWindowFn<>(
-                    originalStrategy.getWindowFn().windowCoder()))
+        Window.<KV<K, V>>into(new IdentityWindowFn<>(originalStrategy.getWindowFn().windowCoder()))
             .triggering(new ReshuffleTrigger<>())
             .discardingFiredPanes()
+            .withOutputTimeFn(OutputTimeFns.outputAtEarliestInputTimestamp())
             .withAllowedLateness(Duration.millis(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()));
 
     return input.apply(rewindow)
-        .apply(GroupByKey.<K, V>create())
+        .apply(ReifyTimestamps.<K, V>inValues())
+        .apply(GroupByKey.<K, TimestampedValue<V>>create())
         // Set the windowing strategy directly, so that it doesn't get counted as the user having
         // set allowed lateness.
         .setWindowingStrategyInternal(originalStrategy)
         .apply("ExpandIterable", ParDo.of(
-            new DoFn<KV<K, Iterable<V>>, KV<K, V>>() {
+            new DoFn<KV<K, Iterable<TimestampedValue<V>>>, KV<K, TimestampedValue<V>>>() {
```

Comment:
I think there's something in the zeitgeist about having a better way to deal with the `V's` of a `KV` without modifying the key; I'm happy to publish a JIRA about this, but I don't know how much utility is present in doing so at the moment (given that the hope is most of the time when the input type is `KV<K, Iterable<V>>` having access to all of the elements is meaningful.